x/service/relay.go

```diff
@@ -1,69 +1,77 @@
 }
 
 func (r Relay) Validate(hostedBlockchains ServiceBlockchains) error {
+	// check to see if the blockchain is empty
 	if r.Blockchain == nil || len(r.Blockchain) == 0 {
 		return EmptyBlockchainError
 	}
+	// check to see if the payload is empty
 	if r.Payload.Data.Bytes() == nil || len(r.Payload.Data.Bytes()) == 0 {
 		return EmptyPayloadDataError
 	}
-	if err := r.ServiceToken.Validate(); err != nil {
-		return errors.New(InvalidTokenError.Error() + " : " + err.Error())
+	// ensure the blockchain is supported
+	if !hostedBlockchains.Contains(hex.EncodeToString(r.Blockchain)) {
+		return UnsupportedBlockchainError
 	}
-	if err := r.IncrementCounter.Validate(r.ServiceToken.AATMessage.ClientPublicKey, r.ServiceToken.Hash()); err != nil {
-		return errors.New(InvalidIncrementCounterError.Error() + " : " + err.Error())
+	// todo check to see if non native blockchain is staked for by the developer
+	// verify that node (self) is of this session
+	if err := SessionSelfVerification(FAKEAPPPUBKEY,
+		r.Blockchain,
+		blockchain.GetLatestSessionBlock().Hash.String());
+		err != nil {
+		return err
 	}
-	if !types.Blockchains(hostedBlockchains).Contains(r.Blockchain.String()) {
-		return UnsupportedBlockchainError
+	// check to see if the service certificate is valid
+	if err := r.ServiceCertificate.Validate(); err != nil {
+		return NewServiceCertificateError(err)
 	}
-	if r.Payload.HttpServicePayload != (HttpServicePayload{}) {
+	if r.Payload.Type() == REST {
 		if len((r.Payload).Method) == 0 {
 			r.Payload.Method = DEFAULTHTTPMETHOD
 		}
 	}
 	return nil
 }
 
-func (r Relay) Execute(hostedBlockchains ServiceBlockchains) (string, error) {
-	if err := r.Validate(hostedBlockchains); err != nil {
-		return "", err
-	}
-	chainURL, err := hostedBlockchains.GetChainURL(r.Blockchain.String())
-	if err != nil {
-		return "", err
-	}
-	result, err := executeHTTPRequest(r.Payload.Data.Bytes(), chainURL, r.Payload.Method)
-	return result, err
+// store the evidence of work done for the relay batch
+func (r Relay) StoreEvidence() error {
+	// grab the relay batch container
+	rbs := GetGlobalRelayBatches()
+	// add the evidence to the proper batch
+	return rbs.AddEvidence(r.ServiceCertificate)
 }
 
-// "executeHTTPRequest" takes in the raw json string and forwards it to the HTTP endpoint
-func executeHTTPRequest(payload []byte, url string, method string) (string, error) {
-	req, err := http.NewRequest(method, url, bytes.NewBuffer(payload))
-	if err != nil {
+// executes the relay on the non-native blockchain specified
+func (r Relay) Execute(hostedBlockchains ServiceBlockchains) (string, error) {
+	if err := r.Validate(hostedBlockchains); err != nil {
 		return "", err
 	}
-	req.Header.Set("Content-Type", "application/json")
-	resp, err := (&http.Client{}).Do(req)
-	if err != nil {
-		return "", err
+	// handle the relay payload based on the type
+	switch r.Payload.Type() {
+	case REST:
```

Comment:
I think these 2 can be encapsulated into HTTP. We need to better define this cause either we need nodes to let Applications know that they are hosting a specific network interface (HTTP vs Websockets vs IPC, etc...).

Reply:
I disagree, in the future there may be different executions based on RPC or REST

Reply:
How is RPC different from REST when all it does at the networking level is sending a POST request to a single endpoint?

Reply:
Good point, will consolidate